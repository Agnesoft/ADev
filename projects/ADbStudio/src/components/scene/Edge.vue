<template>
    <div class="edge" :style="'--top_pos:'+ top_pos+'; --left_pos:'+ left_pos+'; --length:'+ length+'; --angle:'+ angle">
        <div class="content">
            {{ params.id }}
        </div>
    </div>
</template>

<script>
export default {
    name: "Edge",
    props: {
        params: Object,
        viewSize: Object
    },
    computed: {
        top_pos: function(){
            let y = (this.params.coordinates1.y + this.params.coordinates2.y) / 2 * this.viewSize.height;
            return (y).toString()+"px";
        },
        left_pos: function(){
            let x = (this.params.coordinates1.x + this.params.coordinates2.x) / 2 * this.viewSize.width - this.sqrt/2;
            return (x).toString()+"px";
        },
        length: function(){
            return (this.sqrt).toString()+"px";
        },
        angle: function(){
            let a = (this.params.coordinates1.x - this.params.coordinates2.x) * this.viewSize.width;
            let b = (this.params.coordinates1.y - this.params.coordinates2.y) * this.viewSize.height;
            return (Math.PI - Math.atan2(-b, a)).toString()+"rad";
        },
        sqrt: function(){
            let a = (this.params.coordinates1.x - this.params.coordinates2.x) * this.viewSize.width;
            let b = (this.params.coordinates1.y - this.params.coordinates2.y) * this.viewSize.height;
            return Math.sqrt(a * a + b * b);
        } 
    },
}
</script>

<style scoped>
    .edge{
        border: 2px solid var(--secondary-color);
        width:  var(--length); 
        text-align: center;
        height: 0px; 
        transform: rotate(var(--angle));
        position: absolute; 
        top: var(--top_pos); 
        left: var(--left_pos);
        cursor: pointer;
        transition: box-shadow var(--transition-ease); 
    }
    .edge:hover{
        box-shadow:  0 0 3px var(--primary-color), 2px 2px 5px var(--dark-color);
    }
</style>