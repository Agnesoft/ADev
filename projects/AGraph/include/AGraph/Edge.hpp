// Copyright 2020 Michael Vlach
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef AGRAPH_EDGE_HPP
#define AGRAPH_EDGE_HPP

#include "Common.hpp"
#include "Element.hpp"

namespace agraph
{
template<typename GraphType, typename GraphImpl>
class Node;

template<typename GraphType, typename GraphImpl>
class Edge : public Element<Edge<GraphType, GraphImpl>, GraphType>
{
public:
    constexpr Edge() = default;

    [[nodiscard]] constexpr auto from() const -> Node<GraphType, GraphImpl>
    {
        if (this->index() != acore::INVALID_INDEX)
        {
            return this->graph()->node(this->graph()->mData.edge(edgeToIndex(this->index())).from);
        }

        return this->graph()->node(acore::INVALID_INDEX);
    }

    [[nodiscard]] constexpr auto to() const -> Node<GraphType, GraphImpl>
    {
        if (this->index() != acore::INVALID_INDEX)
        {
            return this->graph()->node(this->graph()->mData.edge(edgeToIndex(this->index())).to);
        }

        return this->graph()->node(acore::INVALID_INDEX);
    }

protected:
    constexpr Edge(acore::size_type index, const GraphType *graph) noexcept :
        Element<Edge<GraphType, GraphImpl>, GraphType>{index, graph}
    {
    }

private:
    friend GraphImpl;
};
}

#endif
