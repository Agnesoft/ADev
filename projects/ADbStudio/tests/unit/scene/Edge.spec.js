import { mount  } from '@vue/test-utils';
import Edge from '@/components/scene/Edge.vue';

import { localVue } from '../TestUtils'


describe('Edge',() => {
    let wrapper
    beforeAll(() => {
        wrapper = mount(Edge, {
            localVue,
            propsData: {
                element: {
                    id: -1,
                    coordinates1: {
                        x: 0.2,
                        y: 0.5,
                    },
                    coordinates2: {
                        x: 0.6,
                        y: 0.3,
                    }
                },
                viewSize: {
                    width: 200,
                    height: 100
                }
            }
          });
    })
    it('renders edge id',() => {
        expect(wrapper.text()).toBe("-1");
    })
    
    it('renders edge in the correct position', () => {
        let style = wrapper.attributes().style;
        expect(style).toContain('--left_pos: 38.76');
        expect(style).toContain('--top_pos: 40');
        expect(style).toContain('--length: 82.46');
        expect(style).toContain('--angle: 6.03');
    })
})
