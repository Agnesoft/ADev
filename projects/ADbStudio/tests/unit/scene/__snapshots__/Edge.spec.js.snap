// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edge renders correctly (snapshot) 1`] = `
<div
  class="edge"
  style="--top_pos: 40px; --left_pos: 38.76894374382339px; --length: 82.46211251235322px; --angle: 6.038206644052722rad;"
>
  <!---->
   
  <div
    class="arrow front"
  />
   
  <div
    class="element"
  >
    
    -1

  </div>
</div>
`;
